text/3735-field-projections.md

```diff
@@ -0,0 +1,1541 @@
+    /* built-in macro */
+}
+```
+
+It has the same syntax as the `offset_of!` macro and returns the respective [field type]. It emits an
+error in the following cases:
+
+```rust
+pub mod foo {
+    pub struct Foo {
+        bar: u32,
+        pub baz: i32,
+    }
+
+    // Error: unknown field `barr` of type `Foo`
+    type FooBar = field_of!(Foo, barr);
+}
+
+pub mod bar {
+    // Error: field `bar` of type `Foo` is private
+    type FooBar = field_of!(Foo, bar);
+}
+```
+
+#### `#[projecting]` Attribute
+
+The `#[projecting]` attribute can be put on a struct or union declaration. It requires that the
+type is `#[repr(transparent)]` and that there is a unique field that has non-zero size in general.
+It results in fields of the single non-zero sized type being considered fields of the outer type.
+
+So for example:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Container<T> {
+    inner: T,
+}
+
+struct Foo {
+    bar: i32,
+}
+```
+
+Now `Container<Foo>` has a [field type] associated with `bar` implementing `Field` with:
+- `Base = Container<Foo>`
+- `Type = Container<i32>`
+- `OFFSET = offset_of!(Foo, bar)`
+
+The same is true for the non-generic case:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Container {
+    inner: Foo,
+}
+
+struct Foo {
+    bar: i32,
+}
+```
+
+Here are some error examples:
+
+```rust
+// ERROR: missing `#[repr(transparent)]`
+#[projecting]
+pub struct Container<T> {
+    inner: T,
+}
+
+// ERROR: no field to project onto found, the struct has no fields
+#[projecting]
+#[repr(transparent)]
+pub struct Container {}
+
+// ERROR: no field to project onto found, all fields are always zero-sized
+#[projecting]
+#[repr(transparent)]
+pub struct Container<T> {
+    phantom1: PhantomData<fn() -> T>,
+    phantom2: PhantomData<fn(T)>,
+}
+```
+
+### Field Projection Operator
+
+The field projection operator `->` has the following syntax:
+
+> **<sup>Syntax</sup>**\
+> _ProjectionExpression_ :\
+> &nbsp;&nbsp; [_Expression_] `->` _ProjectionMember_
+>
+> _ProjectionMember_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [IDENTIFIER]
+> &nbsp;&nbsp; | [TUPLE_INDEX]
+
+[IDENTIFIER]: https://doc.rust-lang.org/reference/identifiers.html
+[_Expression_]: https://doc.rust-lang.org/reference/expressions.html
+[TUPLE_INDEX]: https://doc.rust-lang.org/reference/tokens.html#tuple-index
+
+#### Desugaring
+
+```rust
+struct T {
+    field: F,
+}
+
+let t: C<T> = /* ... */;
+let _ = t->field;
+
+// becomes
+
+let _ = Project::<field_of!(<C<T> as Projectable>::Inner, field)>::project(t);
+```
+
+The `C<T>` in the `C<T> as Projectable` comes from a type inference variable over the expression
+`t`.
+
+
+#### `Projectable` & `Project` Traits
+
+Added to `core::ops`:
+
+```rust
+pub trait Projectable: Sized {
+    type Inner: ?Sized;
+}
+
+pub trait Project<F>: Projectable
+where
+    F: Field<Base = Self::Inner>,
+{
+    type Output;
+
+    fn project(self) -> Self::Output;
+}
+```
```

Comment:
does annotating a type `#[projecting]` automatically make it implement the `Projectable` trait? given the desugaring of `->` involves `C<T> as Projectable` it seems it should.

```rust
#[projecting]
struct Container1<T> {
    inner: T,
    zst_stuff: ZstData,
}

// ↓

impl<T> Projectable for Container1<T> {
    type Inner = Self; // ?
}
impl<T> Project<F: Field<Base = Self>> for Container1<T> {
    type Output = F::Type; // ?
    fn project(self) -> Self::Output {
        /* magic */
    }
}
```

Reply:
> @y86-dev Hmm ok. So in a sense `Projectable` is designed to be implemented only for "pointer-like containers" (`&T`, `*T`, `NonNull<T>`, `cell::Ref<T>`) rather than "value-like containers" (`MaybeUninit<T>`, `Cell<T>`).

Yes that is exactly how the two differ.

> This perhaps builds a further case _against_ making `Cow<T>` projectable, since for a `T: Clone` type the owned variant is actually "value-like".

Yeah that is probably true.

Reply:
I added a new section in the guide-level explanation to better convey this:
> Hmm ok. So in a sense Projectable is designed to be implemented only for "pointer-like containers" (&T, *T, NonNull<T>, cell::Ref<T>) rather than "value-like containers" (MaybeUninit<T>, Cell<T>).